src/softfloat.h

```diff
@@ -35,29 +35,44 @@ static inline uint32_t calc_fclass(uint32_t f)
     const uint32_t expn = f & FMASK_EXPN;
     const uint32_t frac = f & FMASK_FRAC;
 
-    /* TODO: optimize with a binary decision tree */
-
     uint32_t out = 0;
+
     /* 0x001    rs1 is -INF */
-    out |= (f == 0xff800000) ? 0x001 : 0;
     /* 0x002    rs1 is negative normal */
-    out |= (expn && (expn != FMASK_EXPN) && sign) ? 0x002 : 0;
     /* 0x004    rs1 is negative subnormal */
-    out |= (!expn && frac && sign) ? 0x004 : 0;
     /* 0x008    rs1 is -0 */
-    out |= (f == 0x80000000) ? 0x008 : 0;
     /* 0x010    rs1 is +0 */
-    out |= (f == 0x00000000) ? 0x010 : 0;
     /* 0x020    rs1 is positive subnormal */
-    out |= (!expn && frac && !sign) ? 0x020 : 0;
     /* 0x040    rs1 is positive normal */
-    out |= (expn && (expn != FMASK_EXPN) && !sign) ? 0x040 : 0;
     /* 0x080    rs1 is +INF */
-    out |= (expn == FMASK_EXPN && !frac && !sign) ? 0x080 : 0;
     /* 0x100    rs1 is a signaling NaN */
-    out |= (expn == FMASK_EXPN && frac && !(frac & FMASK_QNAN)) ? 0x100 : 0;
     /* 0x200    rs1 is a quiet NaN */
-    out |= (expn == FMASK_EXPN && (frac & FMASK_QNAN)) ? 0x200 : 0;
+
+    // Check if it is special value -INF
+    out |= (f == 0xff800000) ? 0x001 : 0;
```

Comment:
`-INF` could be integrated into the if-else statement

Reply:
I rewrote the decision tree as follow, and gained more speedup:
 
```
/* Check the exponent bits */
    if (expn) {
        if (expn != FMASK_EXPN) {
            /* Check if it is negative normal or positive normal */
            out = sign ? 0x002 : 0x040;
        } else {
            /* Check if it is NaN */
            if (frac) {
                out = frac & FMASK_QNAN ? 0x200 : 0x100;
            } else if (!sign) {
                /* Check if it is +INF */
                out = 0x080;
            } else {
                /* Check if it is -INF */
                out = 0x001;
            }
        }
    } else if (frac) {
        /* Check if it is negative or positive subnormal */
        out = sign ? 0x004 : 0x020;
    } else{
        /* Check if it is +0 or -0 */
        out = sign ? 0x008 : 0x010;
    }
 ```

Reply:
@2011eric 
Thanks for your review and suggestions.
In my tests, the current version's execution efficiency is now approximately 69% faster than the original version.

Reply:
Are you able to run arch-test with DEVICE=F?
Just to make sure this implementation is correct.

Reply:
When I attempted to execute the command ```make arch-test RISCV_DEVICE=FZicsr``` as your request, I consistently encountered the following error message, regardless of using the new function or the original one:
```
inferior exit code 0
inferior exit code 0  
ERROR | Segmentation fault (core dumped)
```
I can successfully execute other tests, such as ```RISCV_DEVICE=IMC```, without encountering any errors. I'm not sure about the cause of the error.

Reply:
However, I used a loop to iterate through the entire range of uint32_t values as input and compared the output differences between the original function and the new function. The experimental results showed that for all inputs, both functions would return the same value.

Here is the code used for testing:
```
uint32_t i = 0;
while(1) {
    if(calc_fclass(i) != calc_fclass_new(i)) {
        fprintf(stderr, "  Error:    %x %x %x", i, calc_fclass(i), calc_fclass_new(i));
        return 1;
    }
    if(i == 0xffffffff) {
        break;
    }
    i++;
}
printf("  All tests passed.\n");
return 0;
```